File: AndroidServiceCommunication/app/src/main/java/lib/serviceconnector/BoundService.java
package lib.serviceconnector;

import android.app.Service;
import android.content.Intent;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.widget.Toast;

import static lib.serviceconnector.ServiceManager.MessageReceiver.*;
/**
 * Created by Lenovo7 on 8/23/2015.
 */
public abstract class BoundService extends Service {

    // Target we publish for clients to send messages to IncomingHandler.
    protected final Messenger mReceivingMessenger = new Messenger(new IncomingHandler());

    protected Messenger mSenderMessenger;

    protected abstract void onMessageReceivedFromUI(Message message);

    protected boolean mIsServiceBounded = false;

    @Override
    public void onCreate() {
        super.onCreate();
    }

    private class IncomingHandler extends Handler { // Handler of incoming messages from clients.
        @Override
        public void handleMessage(Message message) {

            if (message.what == SERVICE_CONNECTED) {
                mSenderMessenger = message.replyTo;
            }
            onMessageReceivedFromUI(message);
        }
    }

    /***
     * @param message
     */
    protected final void sendMessageToUI(String message) {
        try {
            mSenderMessenger.send(Message.obtain(null, SERVICE_SENT_DATA, message));
        } catch (RemoteException e) {
            e.printStackTrace();
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    /***
     * @param message
     */
    protected final void sendMessageToUI(Message message) {
        try {
            mSenderMessenger.send(message);
        } catch (RemoteException e) {
            e.printStackTrace();
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        mIsServiceBounded = true;
        Toast.makeText(getApplication(), "onBind", Toast.LENGTH_SHORT).show();
        return mReceivingMessenger.getBinder();
    }

    @Override
    public boolean onUnbind(Intent intent) {
        Toast.makeText(getApplication(), "onUnbind", Toast.LENGTH_SHORT).show();
        mIsServiceBounded = false;
        return super.onUnbind(intent);
    }

    @Override
    public void onDestroy() {
        Toast.makeText(getApplication(), "onDestroy", Toast.LENGTH_SHORT).show();
        super.onDestroy();
    }
}

File: AndroidServiceCommunication/app/src/main/java/lib/connector/BoundService.kt
package lib.connector

import android.app.Service
import android.content.Intent
import android.os.IBinder
import android.os.Message
import android.os.Messenger
import android.os.RemoteException
import android.widget.Toast

/**
 * Service bounded with Activity to send message between activity-service
 * @property receivingMessenger Messenger
 * @property senderMessenger Messenger?
 * @property isServiceBounded Boolean
 */
abstract class BoundService:Service(), HandleMessageCallback {

    // Target we publish for clients to send messages to IncomingHandler.
    private lateinit var receivingMessenger : Messenger

    private var senderMessenger: Messenger?=null

    protected abstract fun onMessageReceivedFromUI(message: Message)

    protected var isServiceBounded = false

    override fun onCreate() {
        super.onCreate()
        receivingMessenger = Messenger(ServiceHandler(this))
    }

    override fun onMessageReceivedFomHandler(message: Message) {
        if (message.what == MessageReceiver.SERVICE_CONNECTED) {
            senderMessenger = message.replyTo
        }
        onMessageReceivedFromUI(message)
    }

    /***
     * @param message
     */
    protected fun sendMessageToUI(message: String) {
        try {
            senderMessenger?.send(Message.obtain(null, MessageReceiver.SERVICE_SENT_DATA, message))
        } catch (e: RemoteException) {
            e.printStackTrace()
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }

    }

    /***
     * @param message
     */
    protected fun sendMessageToUI(message: Message) {
        try {
            senderMessenger?.send(message)
        } catch (e: RemoteException) {
            e.printStackTrace()
        } catch (e: NullPointerException) {
            e.printStackTrace()
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        isServiceBounded = true
        Toast.makeText(application, "onBind", Toast.LENGTH_SHORT).show()
        return receivingMessenger.binder
    }

    override fun onUnbind(intent: Intent): Boolean {
        Toast.makeText(application, "onUnbind", Toast.LENGTH_SHORT).show()
        isServiceBounded = false
        return super.onUnbind(intent)
    }

    override fun onDestroy() {
        Toast.makeText(application, "onDestroy", Toast.LENGTH_SHORT).show()
        super.onDestroy()
    }
}